erigon-lib/recsplit/multiencseq/sequence_reader.go

```diff
@@ -0,0 +1,170 @@
+package multiencseq
+
+import (
+	"github.com/erigontech/erigon-lib/kv/stream"
+	"github.com/erigontech/erigon-lib/recsplit/eliasfano32"
+	"github.com/erigontech/erigon-lib/recsplit/simpleseq"
+)
+
+type EncodingType uint8
+
+const (
+	PlainEliasFano   EncodingType = 0b0
+	SimpleEncoding   EncodingType = 0b10000000
+	RebasedEliasFano EncodingType = 0b10010000
+)
+
+// SequenceReader is used to read serialized number sequences.
+//
+// It is aware of the different encoding types and can read them transparently.
+//
+// This is the "reader" counterpart of SequenceBuilder.
+type SequenceReader struct {
+	currentEnc EncodingType
+	ref        eliasfano32.RebasedEliasFano
+	sseq       simpleseq.SimpleSequence
+}
+
+func ReadMultiEncSeq(baseNum uint64, raw []byte) *SequenceReader {
+	var s SequenceReader
+	s.Reset(baseNum, raw)
+	return &s
+}
+
+// This is a specialized "fast" Count method that shouldn't allocate new objects, but read the count directly
+// from raw data
+func Count(baseNum uint64, data []byte) uint64 {
+	// plain elias fano (legacy)
+	if data[0]&0b10000000 == 0 {
+		return eliasfano32.Count(data)
+	}
+
+	// rebased elias fano
+	if data[0] == 0x90 {
+		return eliasfano32.Count(data[1:])
+	}
+
+	// simple encoding
+	if data[0]&0b11110000 == 0b10000000 {
```

Comment:
use const

Reply:
forgot to use the existing SimpleEncoding const... will also add one for the comparison mask